src/plan_ordered_append.c

```diff
@@ -0,0 +1,188 @@
+#include <nodes/value.h>
+#include <optimizer/paths.h>
+#include <optimizer/pathnode.h>
+#include <optimizer/tlist.h>
+#include <utils/builtins.h>
+#include <utils/typcache.h>
+
+#include "compat.h"
+#include "guc.h"
+#include "hypertable.h"
+#include "hypertable_restrict_info.h"
+#include "constraint_aware_append.h"
+#include "planner.h"
+#include "plan_ordered_append.h"
+
+/*
+ * Check if conditions for doing ordered append optimization are fulfilled
+ */
+bool
+ts_ordered_append_do_optimization(PlannerInfo *root, RelOptInfo *rel, Hypertable *ht, bool *reverse)
+{
+	/* check if optimizations are enabled */
+	if (ts_guc_disable_optimizations || !ts_guc_ordered_append)
+		return false;
+
+	/*
+	 * only do this optimization for hypertables with 1 dimension and queries
+	 * with an ORDER BY and LIMIT clause
+	 */
+	if (ht->space->num_dimensions != 1 || root->parse->sortClause == NIL || root->limit_tuples == -1.0)
+		return false;
+
+	/*
+	 * check that the first element of the ORDER BY clause actually matches
+	 * the first dimension of the hypertable
+	 */
+	{
+		SortGroupClause *sort = linitial(root->parse->sortClause);
+		TargetEntry *tle = get_sortgroupref_tle(sort->tleSortGroupRef, root->parse->targetList);
+		RangeTblEntry *rte = root->simple_rte_array[rel->relid];
+		TypeCacheEntry *tce;
+		char	   *column;
+
+		/* doublecheck rel actually refers to our hypertable */
+		Assert(ht->space->main_table_relid == rte->relid);
+
+		/*
+		 * check the ORDER BY column actually belonging to our hypertable
+		 * Unfortunately resorigtbl is not set for junk columns so we can't
+		 * doublecheck for those this way, but we are only dealing with single
+		 * relations at this level anyway so this should always match
+		 */
+		Assert(tle->resjunk == false || ht->space->main_table_relid != tle->resorigtbl);
+
+		/*
+		 * we only support direct column references for now
+		 */
+		if (!IsA(tle->expr, Var))
+			return false;
+
+		/* check dimension column is our ORDER BY column */
+		column = strVal(list_nth(rte->eref->colnames, AttrNumberGetAttrOffset(castNode(Var, tle->expr)->varattno)));
+		if (namestrcmp(&ht->space->dimensions[0].fd.column_name, column) != 0)
+			return false;
+
+		/*
+		 * check sort operation is either less than or greater than
+		 */
+		tce = lookup_type_cache(castNode(Var, tle->expr)->vartype, TYPECACHE_LT_OPR | TYPECACHE_GT_OPR);
+		if (sort->sortop != tce->lt_opr && sort->sortop != tce->gt_opr)
+			return false;
+
+		if (reverse != NULL)
+			*reverse = sort->sortop == tce->lt_opr ? false : true;
+	}
+
+	if (rel->fdw_private != NULL)
+		((TimescaleDBPrivate *) rel->fdw_private)->appends_ordered = true;
+
+	return true;
+}
+
+/*
+ * this function is very similar to PostgreSQL create_append_path
+ * the main difference is setting pathkeys and setting parallel_aware
+ * to false
+ */
+static AppendPath *
+#if PG96
+create_ordered_append_path(PlannerInfo *root, RelOptInfo *rel, List *subpaths, List *pathkeys, Relids required_outer)
```

Comment:
Since this code is taken from PostgreSQL I think we should include a note about that in top of the file

Reply:
refactored to not pull in postgres code